designs/0034-jacobian.md

```diff
@@ -0,0 +1,176 @@
+
+- $f(y)$ The unconstraining transform
+- $f^{-1}\left(y\right)$ The inverse unconstraining transform
+- $\log |J(y)|$ The log absolute Jacobian determinant function for $f^{-1}$ chosen such that the resulting distribution over the constrained variables is uniform
+- $V(y)$ that tests that $x \in \mathcal{X}$ (i.e., that $x$ satisfies the constraint defining $\mathcal{X}$).
+
+Having the Stan language define types for transforms from the unconstrained to constrained space allows users to focus on modeling in the constrained space while algorithm developers focus on developing algorithms in the unconstrained space. 
+This is very nice for both parties since it is both of their preferred spaces to work in.
+
+Most of the time this encapsulation is very good. 
+The main issue is that Stan users either have to write code that either only uses Stan's built-in transforms to stay in the constrained space or have to always have the jacobian calcuations enabled while working in the constrained space.
+
+
+
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The following section is a draft of the docs for the new constraint and `jacobian` keyword.
+
+### Using the jacobian Keyword
+
+The `jacobian` keyword is introduced within the `transformed parameters` block and the functions ending in `_jacobian` to allow the Jacobian accumulator to be incremented by the log absolute determinant of the Jacobian of the constraining transform. 
+This keyword behaves similarly to the `target` keyword, allowing users to account for the change of variables in the probability density function. 
+The key difference between the two is that `jacobian` will only be accumulated into for algorithms that request it. Algorithms such as maximum likelihood would request the jacobian to not be accumulated. 
+The `jacobian` keyword is available only in the `transformed parameters` block and within functions ending in `_jacobian`.
+
+
+### Example Usage
+
+The following stan program defines an `upper_bound_jacobian` function for constrainting `real` and `vector` types.
+
+```stan
+functions {
+real upper_bound_jacobian(real x, real ub) {
+  jacobian += x;
+  return upper_bound - exp(x);
+}
+vector upper_bound_jacobian(vector x, real ub) {
+  jacobian += x;
+  return upper_bound - exp(x);
+}
+}
+
+data {
+ real ub;
+ int N;
+}
+
+parameters {
+  real b_raw;
+  vector[N] b_vec_raw;
+}
+
+transformed parameters {
+  real b = upper_bound_jacobian(b_raw, ub)
+  vector[N] b = upper_bound_jacobian(b_vec_raw, ub)
+}
+```
+
+The following stan program calculates the same constraint and jacobian directly in the transformed parameters block.
+
+```stan
+data {
+ real ub;
+ int N;
+}
+
+parameters {
+  real b_raw;
+}
+
+transformed parameters {
+  jacobian += b_raw;
+  real b = upper_bound - exp(b_raw);
+}
+```
+
+
+# Reference-level explanation
```

Comment:
You removed the section on what this compiles to in C++, but I think bringing a version of it back would be nice, even if it is hopefully pretty clear already to people who have worked with Stan's C++